public/jira/raw/rollback/rollback.ts

```diff
@@ -0,0 +1,311 @@
+/**
+ * This module is responsible for rolling back Jira issues to their previous states based on their changelogs.
+ * It processes changes to fields such as Sprint, Fix Versions, Parent associations, and Status,
+ * ensuring that issues are accurately reverted to their historical configurations.
+ */
+import { parseDateISOString } from "../../../date-helpers";
+import {
+  Change,
+  FixVersion,
+  JiraIssue,
+  Sprint,
+  Status,
+} from "../../shared/types";
+
+interface RolledBackMetadata {
+  rolledbackTo: Date;
+  didNotExist?: boolean;
+}
+
+interface RolledBackJiraIssue extends JiraIssue {
+  rollbackMetadata: RolledBackMetadata;
+}
+
+interface FieldLookup<T> {
+  ids: Map<string | number, T>;
+  names: Map<string, T>;
+}
+
+interface RollbackLookupData {
+  sprints: FieldLookup<Sprint>;
+  versions: FieldLookup<FixVersion>;
+  statuses: FieldLookup<Status>;
+}
+
+function getSprintNumbers(value: string): number[] | null {
+  if (value === "") {
+    return null;
+  } else {
+    return value.split(",").map((num) => +num);
+  }
+}
+
+function getSprintNames(value: string): string[] | null {
+  if (value === "") {
+    return null;
+  } else {
+    return value.split(",").map((name) => name.trim());
+  }
+}
+
+type FieldRollbackFunction = (
+  lastReturnValue: unknown,
+  change: Change,
+  fieldName: string,
+  data: RollbackLookupData
+) => { [key: string]: unknown };
+
+type FieldRollbackFunctionMap = { [fieldName: string]: FieldRollbackFunction };
+
+export const fields: FieldRollbackFunctionMap = {
+  // from will look like "1619, 1647"
+  // we need to update `lastReturnValue` to have
+  // only the right sprints
+  Sprint: function (lastReturnValue, change, fieldName, { sprints }) {
+    const sprintNumbers = getSprintNumbers(change.from || "");
+    const sprintNames = getSprintNames(change.fromString || "");
+
+    if (sprintNumbers === null) {
+      return { [fieldName]: null };
+    } else {
+      return {
+        [fieldName]: sprintNumbers
+          .map((number, i) => {
+            // REMOVE IN PROD
+            if (sprints.ids.has(number)) {
+              return sprints.ids.get(number);
+            } else if (sprints.names.has(sprintNames![i])) {
+              return sprints.names.get(sprintNames![i]);
+            } else {
+              // TODO: change to async so we can go request all of these
+              console.warn("Can't find sprint ", number, sprintNames![i]);
+              return undefined;
+            }
+          })
+          .filter((x): x is Sprint => x !== undefined),
```

Comment:
what version of ts are we on, in the new ones you don't need the predicate :D

Reply:
Yeah you are correct, I recently needed to add a type guard on my previous project because it was not narrowing down the type (maybe it was an if condition, cant remember :D)

Reply:
Removed type guard